apps/frontend/src/lib/storage/localStorage/user/connection/message/index.ts

```diff
@@ -138,14 +98,137 @@ export const processNewMessages = async (
+    // Double hash the signature private key, use as identifier, use single hash version as revocation code
+    tapReceiverHash = sha256(sha256(user.signaturePrivateKey) || DEVCON);
+  }
+
+  const newBackupEntries: CreateBackupData[] = [];
+
+  for (const message of messages) {
+
+    const sendHash: boolean = user.tapGraphEnabled === true;
+
+    // Make backup, add edge to user object
+    const backup: CreateBackupData = createEdgeBackup({
+      email: user.email,
+      password: session.backupMasterPassword,
+
+      // Arguments for backup
+      edgeId: message.edge.edgeId,
+      connectionUsername: message.edge.connectionUsername, // The user that tapped / sent the message
+      sentHash: sendHash, // whether *you* sent the hash or not
+      timestamp: message.edge.timestamp
+    });
+
+    // Push onto array
+    newBackupEntries.push(backup);
+
+    // If setting is not enabled, no need to push hash to row
+    if (!sendHash) {
+      continue;
+    }
+
+    try {
+      // Upsert row, returns edge ID
+      await upsertSocialGraphEdge(session.authTokenValue, message.edge.edgeId, null, tapReceiverHash);
+    } catch (error) {
+      console.error(`Error while upserting edge: ${errorToString(error)}`);
+      return newBackupEntries;
+    }
+  }
+
+  return newBackupEntries;
+}
+
+
+export const processNewMessages = async (
+  messages: MessageData[]
+): Promise<void> => {
+  if (messages.length === 0) {
+    return;
+  }
+
+  const { user, session } = await getUserAndSession();
+
+  const connectionTapBacks: Record<string, TapBackMessage[]> = {};
+  const edgeMessages: EdgeMessage[] = [];
+
+  let lastMessageTimestamp: Date | null = null;
+  const sortedMessages = messages.sort(
+    (a, b) => a.createdAt.getTime() - b.createdAt.getTime()
+  );
+  for (const message of sortedMessages) {
+    // Ensure messages are in order and update lastMessageTimestamp
+    if (lastMessageTimestamp && message.createdAt < lastMessageTimestamp) {
+      console.error("Messages are not in order");
+      continue;
+    }
+    lastMessageTimestamp = message.createdAt;
+
+    try {
+      const { serializedData, senderSignaturePublicKey } = await decryptReceivedMessage({
+          messageData: message,
+          encryptionPrivateKey: user.encryptionPrivateKey,
+        })
+
+      if (serializedData.match(MessageType.TAP_BACK.toString())) {
+        const tapBackMessage = parseSerializedTapBackMessage(serializedData);
+        if (Object.keys(connectionTapBacks).includes(senderSignaturePublicKey)) {
+          connectionTapBacks[senderSignaturePublicKey] = [
+            ...connectionTapBacks[senderSignaturePublicKey],
+            tapBackMessage,
+          ];
+        } else {
+          connectionTapBacks[senderSignaturePublicKey] = [tapBackMessage];
+        }
+        continue;
+      } else if (serializedData.match(MessageType.TAP_GRAPH_EDGE.toString())) {
+        const edgeMessage = parseSerializedEdgeMessage(serializedData);
```

Comment:
as a future note, here we should check that the `connectionUsername` in the edge message matches the `signaturePublicKey`. this is not an easy fix in the current setup since at this point the connection struct might not exist, but would be doable with a public registry

Reply:
signaturePublicKey in your connections object?